### PR TITLE
use param_env on the trait_cache key

### PR DESCRIPTION
**Pull request by qmx**
We bailed from making trans_fulfill_obligation return `Option` or `Result`, just made it less prone to crashing outside trans

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/traits/trans/mod.rs [hunk 1]
```diff
@@ -33,22 +33,22 @@ impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
     /// obligations *could be* resolved if we wanted to.
```
Comment (nikomatsakis) [inline src/librustc/traits/trans/mod.rs]: Nit: Can we extend this comment to say "/// Assumes that this is run after the entire crate has been successfully type-checked."

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Just a nit.

**Comment by kennytm** [comment]
Needs to submit a PR to miri, and set miri's [toolstate](https://github.com/rust-lang/rust/blob/master/src/tools/toolstate.toml) to "Broken". (cc @oli-obk @RalfJung @eddyb)

```
[00:47:34] error[E0061]: this function takes 3 parameters but 2 parameters were supplied
[00:47:34]     --> /checkout/src/tools/miri/src/librustc_mir/interpret/eval_context.rs:2419:45
[00:47:34]      |
[00:47:34] 2419 |     let vtbl = tcx.trans_fulfill_obligation(DUMMY_SP, ty::Binder(trait_ref));
[00:47:34]      |                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 parameters
[00:47:34] 
[00:47:35] error: aborting due to previous error
[00:47:35] 
[00:47:35] error: Could not compile `rustc_miri`.
```

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**